File: frontend/src/serversStore.ts
import { defineStore } from "solidjs-storex";
import { GetServersJSON, SaveServersJSON } from '../bindings/github.com/andreitelteu/phpmyadmin-desktop/app';
import { produce } from "solid-js/store";
import { createEffect } from "solid-js";

type Server = {
    id: string,
    name: string,
    host: string,
    port: number,
    username: string,
    password: string,
    tunnel: {
        enabled: boolean,
        host: string,
        port: number,
        username: string,
        password: string,
        authMethod: 'password' | 'publicKey',
        privateKey: string,
        passphrase: string,
    },
}
type ServersState = {
    list: Server[],
}

const store = defineStore({
    state: { list: []} as ServersState,
    options: {
        persistent: false,
    },
    actions: (state, set) => ({
        set(...data: any[]) {
            // @ts-expect-error
            set(...data);
            console.log('set', data);
        },
        findById(id: string) {
            return state.list.find(s => s.id === id);
        },
        newServer() {
            set(produce(s => {
                s.list.push({
                    // New connections must have a stable ID before their first
                    // autosave. Existing blank legacy IDs are intentionally not
                    // backfilled: the user can delete and recreate those rows.
                    id: crypto.randomUUID(),
                    name: '',
                    host: '',
                    port: 3306,
                    username: '',
                    password: '',
                    tunnel: {
                        enabled: false,
                        host: '',
                        port: 22,
                        username: '',
                        password: '',
                        authMethod: 'password',
                        privateKey: '',
                        passphrase: '',
                    },
                })
            }))
        },
        removeServer(index: number) {
            set(produce(s => {
                if (index >= 0 && index < s.list.length) {
                    s.list.splice(index, 1);
                }
            }))
        },
        updateServer(index: number, data: Partial<Server>) {
            set(produce(s => {
                if (data.tunnel !== undefined) {
                    data.tunnel = Object.assign(s.list[index].tunnel, data.tunnel);
                }
                Object.assign(s.list[index], data);
            }))
        },
    }),
});
export default store;

GetServersJSON().then(res => {
    try {
        let data = JSON.parse(res);
        store()[1].set(data);
    } catch (e) {
        console.error(e);
    }
})
.finally(() => {
    createEffect(() => {
        const data = store()[0];
        SaveServersJSON(JSON.stringify(data, null, 2)).catch(console.error);
    });
});

File: serversStore.go
package main

import (
	"encoding/json"
	"fmt"

	wailsconfigstore "github.com/AndreiTelteu/wails-configstore"
)

// TunnelConfig describes an optional local SSH port-forward for a database
// server. It is persisted alongside the connection definition.
type TunnelConfig struct {
	Enabled    bool   `json:"enabled"`
	Host       string `json:"host"`
	Port       int    `json:"port"`
	Username   string `json:"username"`
	Password   string `json:"password"`
	AuthMethod string `json:"authMethod"`
	PrivateKey string `json:"privateKey"`
	Passphrase string `json:"passphrase"`
}

type ServerConfig struct {
	ID       string       `json:"id"`
	Name     string       `json:"name"`
	Host     string       `json:"host"`
	Port     int          `json:"port"`
	Username string       `json:"username"`
	Password string       `json:"password"`
	Tunnel   TunnelConfig `json:"tunnel"`
}

type ServersConfig struct {
	List []ServerConfig `json:"list"`
}

func (s *ServersConfig) FindByID(id string) *ServerConfig {
	for index := range s.List {
		if s.List[index].ID == id {
			return &s.List[index]
		}
	}
	return nil
}

func GetServersConfig(conf *wailsconfigstore.ConfigStore) (*ServersConfig, error) {
	data, err := conf.Get("servers.json", `{"list":[]}`)
	if err != nil {
		return nil, fmt.Errorf("read servers configuration: %w", err)
	}

	var serversConfig ServersConfig
	if err := json.Unmarshal([]byte(data), &serversConfig); err != nil {
		return nil, fmt.Errorf("parse servers configuration: %w", err)
	}
	return &serversConfig, nil
}
